rero_ils/modules/documents/serializers/base.py

```diff
@@ -0,0 +1,257 @@
+# -*- coding: utf-8 -*-
+#
+# RERO ILS
+# Copyright (C) 2022 RERO
+# Copyright (C) 2022 UCLouvain
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as published by
+# the Free Software Foundation, version 3 of the License.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+"""Mixin helper class."""
+
+from rero_ils.modules.contributions.utils import \
+    get_contribution_localized_value
+
+CREATOR_ROLES = [
+    'aut', 'cmp', 'cre', 'dub', 'pht', 'ape', 'aqt', 'arc', 'art', 'aus',
+    'chr', 'cll', 'com', 'drt', 'dsr', 'enj', 'fmk', 'inv', 'ive', 'ivr',
+    'lbt', 'lsa', 'lyr', 'pra', 'prg', 'rsp', 'scl'
+]
+
+
+class BaseDocumentFormatterMixin:
+    """Base formatter class."""
+
+    def __init__(self, record):
+        """Initialize RIS formatter with the specific record."""
+        self.record = record
+
+    def format(self):
+        """Return RIS export for single record."""
+        raise NotImplementedError
+
+    def _get_document_types(self):
+        """Return document types."""
+        doc_types = []
+        for main_type, subtype in [
+            (doc_type.get('main_type'), doc_type.get('subtype'))
+                for doc_type in self.record['type']]:
+
+            if subtype:
+                main_type = f'{main_type} / {subtype}'
+            doc_types.append(main_type)
+
+        return doc_types
+
+    def _get_pid(self):
+        """Return reference id."""
+        return self.record['pid']
+
+    def _get_title(self):
+        """Return first title."""
+        return next(
+            filter(lambda x: x.get('type') == 'bf:Title',
+                   self.record.get('title'))
+        ).get('_text')
```

Comment:
add a default empty dict value to `next` to not raise an exception if a document doesn't have any title (It should be possible with `_draft:true` attribute that skip any JSON schema validation).

```next(filter(...), {}).get('_text')```